CellTowerMarker: Show PSC and RAT

SecUpwN opened the pull request:
This pull request shall fix #723 and enable the user to see `PSC` and `RAT` when clicking a marked cell.
Need help here. @f3ndot what is missing in our code so that these things show in a clicked CellTower?

SecUpwN committed:
Uncommented PSC and RAT

f3ndot commented:
MarkerData.java is missing the setters, getters, and class properties for RAT and PSC. See patch file to make it available (not you'll need to change the instantiation of the class in all place to pass in RAT and PSC now)

```diff
diff --git a/AIMSICD/src/main/java/com/SecUpwN/AIMSICD/map/MarkerData.java b/AIMSICD/src/main/java/com/SecUpwN/AIMSICD/map/MarkerData.java
index 42ee95d..ecedde7 100644
--- a/AIMSICD/src/main/java/com/SecUpwN/AIMSICD/map/MarkerData.java
+++ b/AIMSICD/src/main/java/com/SecUpwN/AIMSICD/map/MarkerData.java
@@ -19,9 +19,8 @@ public class MarkerData {
     public final String lac;        // LAC
     private final String mcc;       // remove and use PC: MCC+MNC
     private final String mnc;       // remove and use PC: MCC+MNC
-    // todo: Add PSC and RAT
-    //private final String psc;     // PSC
-    //private final String rat;     // RAT
+    private final String psc;       // PSC
+    private final String rat;       // RAT

     private final String samples;   // samples
     public final boolean openCellID; // ??
@@ -33,6 +32,8 @@ public class MarkerData {
                String local_area_code,
                String mobile_country_code,
                String mobile_network_code,
+               String the_psc,
+               String the_rat,
                String samples_taken,
                boolean openCellID_Data) {
         cellID = cell_id;
@@ -41,6 +42,8 @@ public class MarkerData {
         lac = local_area_code;
         mcc = mobile_country_code;
         mnc = mobile_network_code;
+        psc = the_psc;
+        rat = the_rat;
         samples = samples_taken;
         openCellID = openCellID_Data;
     }
@@ -66,6 +69,20 @@ public class MarkerData {
         return ("00" + mnc).substring(mnc.length());
     }

+    public String getPSC() {
+        if (psc == null || psc.isEmpty()) {
+            return "N/A";
+        }
+        return psc;
+    }
+
+    public String getRAT() {
+        if (rat == null || rat.isEmpty()) {
+            return "N/A";
+        }
+        return rat;
+    }
+
     // (Mobile Network Operator) Provider Code in form: MCC-MNC
     public String getPC() {
         return getMCC() +"-"+ getMNC();
```

f3ndot commented:
Also `R.id.rat` is missing

SecUpwN commented:
>See patch file to make it available (note you'll need to change the instantiation of the class in all place to pass in RAT and PSC now)

Thanks for providing your patch file, @f3ndot! I am not on my own computer right now, would you please do the effort and push your corrected patch into branch `CellTowerMarker` so that this gets completed?

f3ndot commented:
I may require collaborator rights on the repo, otherwise I'll have to make
a PR against the branch via my personal fork

On Fri, Feb 26, 2016, 5:17 PM Security: PWNED <notifications@github.com>
wrote:

> See patch file to make it available (note you'll need to change the
> instantiation of the class in all place to pass in RAT and PSC now)
>
> Thanks for providing your patch file, @f3ndot <https://github.com/f3ndot>!
> I am not on my own computer right now, would you please do the effort and
> push your corrected patch into branch CellTowerMarker so that this gets
> completed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/pull/772#issuecomment-189505729>
> .
>

SecUpwN commented:
>I may require collaborator rights on the repo, otherwise I'll have to make a PR against the branch via my personal fork

Ha, neat way to ask for elevation of your rights here! :) Access granted.

f3ndot committed:
Add PSC to map marker info window w/ i18n if missing

Noticed that if OCID doesn't contain a PSC value for cells, the app sets PSC to
`666` in the database. Most places will display "invalid" but that's not
translateable.

Refactored the validation method to Cell as static and use it to display the
i18n'd string "Unknown" which I feel best conveys when a PSC would be out-of-
bounds.

Future work should translate unknown string for all langs and use it instead of
the `invalid` string literal.

f3ndot commented:
Things I'm still working on:

- [ ] Translate the new `unknown` String key to the best of my ability
- [ ] Make sure all PSC values displayed to the user are translated if `invalid` is shown
- [ ] Repeat for RAT

f3ndot commented on AIMSICD/src/main/java/com/SecUpwN/AIMSICD/adapters/AIMSICDDbAdapter.java:
```diff
@@ -566,8 +566,9 @@ public boolean populateDBeImport() {
+                        // Some OCID data may not contain PSC so we indicate this with an out-of-range
+                        // PSC value. Should be -1 but hey people already imported so we're stuck with
+                        // this.
                         int iPsc = 666;
```
Had a good chuckle when I found this little tidbit.

All of the OCID cells in my area have missing PSC, so everything was showing up as "666" in the map :laughing:

f3ndot replied:
Because I refactored `validatePscValue` to check lower valid bounds too, we *could* change to `-1` but I'd rather be consistent with the dataset already on people's phones.

f3ndot committed:
Translate all validated uses of PSC

PSC isn't always available, so we translate "Unknown" if absent

f3ndot committed:
Add human-readable RAT/Network Type to map marker details

Also turns RAT/Network Type integer in BTA Measurments DB viewer into
human-readable text as well.

This commit, in my opinion, is a little hax.

There should be a refactor of the app to unify the terms of RAT/Network Type
and how its stored/represented/displayed at what layers.

f3ndot committed:
Add "Unknown" translations

f3ndot committed:
Adhere to JLS for static keyword order

SecUpwN commented:
Thanks for digging deep into our code to fix as many Issues as possible, @f3ndot. You'll likely discover even more strange tidbits. Please do me the favor and either *remove* or move unuseful comments, TODO and possibly even CHANGELOGS into a fresh open Issue (if a file still needs editing if it contains something like TODO notes). We need to clean our code as best as possible for better development.

SecUpwN commented:
@f3ndot, let me know here when this is ready to merge. You're doing an awesome job!

f3ndot commented:
@SecUpwN So it's not the most glorious code but it's functional and helped in other parts along the way.

If you're ok with it, let's merge it and a future effort to better refactor the code can be undertaken by myself or some new fresh-faced developer :wink:

SecUpwN commented:
>If you're ok with it, let's merge it and a future effort to better refactor the code can be undertaken by myself or some new fresh-faced developer :wink:

Fair thing, feel free to improve your own code later on. We have so few developers that is difficult to expect that someone else will be take another closer look at it after it has been merged when it works.

And since you're doing such a great job, I've added you to our Buildozer account, which enables you to [grab the very bleeding edge](https://play.google.com/store/apps/details?id=io.buildozer.app) of our APK which gets compiled once *anything* gets merged here. Enjoy!

SecUpwN commented:
Approved!
